src/dynamic_action.rs

```diff
@@ -0,0 +1,142 @@
+//! The dynamic action type.
+//!
+//! This type is primarily intended for use when declaring every single action in a single enum isn't feasible,
+//! for example due to a complex crate hierarchy where various feature crates want to individually declare their own
+//! actions without needing to coordinate with other crates.
+//!
+//! Example:
+//! ```
+//! # use leafwing_input_manager::dynamic_action::{DynActionMarker, DynActionRegistry};
+//!
+//! // In crate `feature_one`
+//! #[derive(DynActionMarker)]
+//! struct FeatureOneAction;
+//!
+//! fn register_feature_one_actions(registry: &mut DynActionRegistry) {
+//!     // There could potentially be many types registered here
+//!     registry.register::<FeatureOneAction>();
+//! }
+//!
+//! // In crate `feature_two`
+//! #[derive(DynActionMarker)]
+//! struct FeatureTwoAction;
+//!
+//! fn register_feature_two_actions(registry: &mut DynActionRegistry) {
+//!     // There could potentially be many types registered here
+//!     registry.register::<FeatureTwoAction>();
+//! }
+//!
+//! // In crate `top_level_crate`, which depends on `feature_one` and `feature_two`
+//!
+//! let mut registry = DynActionRegistry::get().unwrap();
+//! register_feature_one_actions(&mut registry);
+//! register_feature_two_actions(&mut registry);
+//! registry.finish();
+//!
+//! ```
+
+use std::any::TypeId;
+
+use bevy::{
+    prelude::{App, Resource},
+    utils::HashMap,
+};
+use once_cell::sync::OnceCell;
+
+use crate::Actionlike;
+
+pub use leafwing_input_manager_macros::DynActionMarker;
+
+static DYN_ACTION_MAP: OnceCell<HashMap<TypeId, usize>> = OnceCell::new();
```

Comment:
Comments please for this strategy: I can follow what it does but it's definitely going to be rough for new contributors.